pkg/api/bzz.go

```diff
@@ -0,0 +1,99 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package api
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"hash"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"strings"
+
+	bmtlegacy "github.com/ethersphere/bmt/legacy"
+	"github.com/gorilla/mux"
+	"golang.org/x/crypto/sha3"
+
+	"github.com/ethersphere/bee/pkg/jsonhttp"
+	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/swarm"
+)
+
+func hashFunc() hash.Hash {
+	return sha3.NewLegacyKeccak256()
+}
+
+func (s *server) bzzUploadHandler(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	data, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		s.Logger.Debugf("bzz: read error: %v", err)
+		s.Logger.Error("bzz: read error")
+		jsonhttp.InternalServerError(w, "cannot read request")
+		return
+	}
+
+	p := bmtlegacy.NewTreePool(hashFunc, swarm.SectionSize, bmtlegacy.PoolSize)
+	hasher := bmtlegacy.New(p)
+	span := len(data)
+	b := make([]byte, 8)
+	binary.LittleEndian.PutUint64(b, uint64(span))
+	data = append(b, data...)
+	err = hasher.SetSpan(int64(span))
+	if err != nil {
+		s.Logger.Debugf("bzz: hasher set span: %v", err)
+		s.Logger.Error("bzz: hash data error")
+		jsonhttp.InternalServerError(w, "cannot hash data")
+		return
+	}
+	_, err = hasher.Write(data[8:])
+	if err != nil {
+		s.Logger.Debugf("bzz: hasher write: %v", err)
+		s.Logger.Error("bzz: hash data error")
+		jsonhttp.InternalServerError(w, "cannot hash data")
+		return
+	}
+	addr := swarm.NewAddress(hasher.Sum(nil))
+	_, err = s.Storer.Put(ctx, storage.ModePutUpload, swarm.NewChunk(addr, data[8:]))
+	if err != nil {
+		s.Logger.Debugf("bzz: write error: %v, addr %s", err, addr)
+		s.Logger.Error("bzz: write error")
+		jsonhttp.InternalServerError(w, "write error")
+		return
+	}
+	w.Header().Set("Content-Type", "text/plain")
+	_, _ = io.Copy(w, strings.NewReader(addr.String()))
```

Comment:
I think that we should have a json response here, since all other responses are json encoded. We could add check on the accept request header but I am of the opinion to keep to json as default, except where we need to return application/octet-stream.